players/base.py

```diff
@@ -0,0 +1,197 @@
+import xbox
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class Player():
+    def __init__(self, name: str, id: int, battle_menu: List[int]):
+        self.name = name
+        self.id = id
+        self.struct_offset = id * 94
+        self.char_rng = 20 + id
+        self.battle_menu = battle_menu
+        
+
+    def __eq__(self, other):
+        if isinstance(other, int):
+            return self.id == other
+        else:
+            return self.id == other.id
+        
+        
+    def __str__(self) -> str:
+        return self.name
+
+
+    def _read_char_offset_address(self, address):
+        return memory.main.read_val(address + self.struct_offset)
+        
+
+    def _read_char_battle_offset_address(self, address, offset):
+        return memory.main.read_val(address + (90 * offset))
+        
+        
+    def _read_char_battle_state_address(self, offset):
+        pointer = memory.main.read_val(0x00D334CC, 4)
+        return memory.main.read_val(pointer + offset + (0xF90 * self.id), 1)
+
+        
+    def _read_char_stat_offset_address(self, address):
+        pointer = memory.main.read_val(0x003AB9B0, 4)
+        return memory.main.read_val(pointer + address + (0x94 * self.id), 1)
+        
+    def navigate_to_battle_menu(self, target):
+        """Different characters have different menu orders."""
+        current_position = memory.main.battle_menu_cursor()
+        target_position = self.battle_menu.index(target)
+        while current_position != target:
+            if self.battle_menu.index(current_position) > target_position:
+                xbox.tap_up()
+            else:
+                xbox.tap_down()
+            current_position = memory.main.battle_menu_cursor()
+        
+        
+    def luck(self) -> int:
+        return self._read_char_stat_offset_address(0x34)
+
+        
+    def accuracy(self) -> int:
+        return self._read_char_stat_offset_address(0x36)
+        
+        
+    def affection(self) -> int:
+        if self.id == 0:
+            return 255
+        return memory.main.read_val(0x00D2CABC + (4 * self.id), 1)
+        
+        
+    def next_crits(self, enemy_luck: int, length: int = 20) -> List[int]:
+        """Note that this says the number of increments, so the previous roll will be a hit, and this one will be the crit."""
+        results = []
+        cur_rng = memory.main.rng_from_index(self.char_rng)
+        cur_rng = memory.main.roll_next_rng(cur_rng, self.char_rng)
+        cur_rng = memory.main.roll_next_rng(cur_rng, self.char_rng)
+        index = 2
+        while len(results) < length:            
+            crit_roll = memory.main.s32(cur_rng & 0x7FFFFFFF ) % 101
+            crit_chance = self.luck() - enemy_luck
+            if crit_roll < crit_chance:
+                results.append(index)
+                index += 1
+        return results
+        
+
+    def next_crit(self, enemy_luck) -> int:
+        return self.next_crits[0]
```

Comment:
Is this valid? next_crits is a function that returns a list, not a list. Or is this some neat python that actually works?

Reply:
Nope, just a typo bug.